randovania/bitpacking/construct_pack.py

```diff
@@ -0,0 +1,200 @@
+            field_name: obj[field_name]
+            for field_name in self.field_types.keys()
+        })
+
+    def _encode(self, obj: T, context, path) -> construct.Container:
+        return construct.Container(
+            (field_name, getattr(obj, field_name))
+            for field_name in self.field_types.keys()
+        )
+
+
+def _construct_dataclass(cls: type[T]) -> ConstructTypedStruct:
+    resolved_types = typing.get_type_hints(cls)
+    field_types: dict[str, type] = {
+        field.name: resolved_types[field.name]
+        for field in dataclasses.fields(cls)
+        if field.init
+    }
+    return ConstructTypedStruct(cls, field_types)
+
+
+def _construct_for_named_tuple(cls) -> construct.Construct:
+    resolved_types = typing.get_type_hints(cls)
+
+    field_types: dict[str, type] = {
+        field_name: resolved_types[field_name]
+        for field_name in cls._fields
+    }
+
+    return ConstructTypedStruct(cls, field_types)
+
+
+class UUIDAdapter(construct.Adapter):
+    def __init__(self):
+        super().__init__(construct.Bytes(16))
+
+    def _decode(self, obj: bytes, context, path) -> uuid.UUID:
+        return uuid.UUID(bytes=obj)
+
+    def _encode(self, obj: uuid.UUID, context, path) -> bytes:
+        return obj.bytes
+
+
+class DatetimeAdapter(construct.Adapter):
+    def __init__(self):
+        super().__init__(construct.ZigZag)
+
+    def _decode(self, obj: int, context, path) -> datetime.datetime:
+        return datetime.datetime.fromtimestamp(obj / 10000000, tz=datetime.timezone.utc)
+
+    def _encode(self, obj: datetime.datetime, context, path) -> int:
+        if obj.tzinfo != datetime.timezone.utc:
+            raise construct.ConstructError(f"{obj} is not an UTC datetime", path)
+
+        return int(obj.timestamp() * 10000000)
+
+
+_direct_mapping = {
+    bool: construct.Flag,
+    int: construct.ZigZag,
+    str: BinStr,
+    uuid.UUID: UUIDAdapter(),
+    datetime.datetime: DatetimeAdapter(),
+}
+
+
+def _tuple_unwrap(con: construct.Construct):
+    return construct.ExprAdapter(
+        con,
+        decoder=lambda obj, ctx: tuple(obj),
+        encoder=lambda obj, ctx: obj,
+    )
+
+
+@functools.cache
+def construct_for_type(type_: type) -> construct.Construct:
+    type_, is_optional = type_lib.resolve_optional(type_)
+
+    if is_optional:
+        return construct.ExprAdapter(
+            construct.PrefixedArray(construct.Byte, construct_for_type(type_)),
+            decoder=lambda obj, ctx: obj[0] if obj else None,
+            encoder=lambda obj, ctx: [obj] if obj is not None else [],
+        )
+
+    type_origin = typing.get_origin(type_)
+
+    if type_ in _direct_mapping:
+        return _direct_mapping[type_]
+
+    if issubclass(type_, Enum):
+        enum_arr = list(type_)
+        return construct.ExprAdapter(
+            construct.VarInt,
+            decoder=lambda obj, ctx: enum_arr[obj],
+            encoder=lambda obj, ctx: enum_arr.index(obj),
+        )
```

Comment:
how safe is it to depend on the exact order of values in an enum? I don't think we change them that frequently but I figured I'd ask

Reply:
For anything that uses construct dataclass, my assumption is that both need to use the exact same version.
Adding am enum value can break, or reordering, but the same applies to changing the dataclass